sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/OperatorsShouldBeOverloadedConsistently.cs

```diff
@@ -0,0 +1,132 @@
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class OperatorsShouldBeOverloadedConsistently : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4050";
+        private const string MessageFormat = "Provide an implementation for: {0}.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static class MethodName
+        {
+            public const string OperatorPlus = "operator+";
+            public const string OperatorMinus = "operator-";
+            public const string OperatorEquals = "operator==";
+            public const string OperatorNotEquals = "operator!=";
+
+            public const string ObjectEquals = "Object.Equals";
+            public const string ObjectGetHashCode = "Object.GetHashCode";
+        }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+
+                if (classSymbol == null || classDeclaration.Identifier.IsMissing)
+                {
+                    return;
+                }
+
+                var classMethods = classSymbol
+                    .GetMembers()
+                    .OfType<IMethodSymbol>()
+                    .Where(m => !m.IsConstructor())
+                    .ToList();
+
+                var implementedMethods = new HashSet<string>();
+                implementedMethods.Add(GetNameOrDefault(MethodName.OperatorPlus, classMethods, KnownMethods.IsOperatorBinaryPlus));
+                implementedMethods.Add(GetNameOrDefault(MethodName.OperatorMinus, classMethods, KnownMethods.IsOperatorBinaryMinus));
+                implementedMethods.Add(GetNameOrDefault(MethodName.OperatorEquals, classMethods, KnownMethods.IsOperatorEquals));
+                implementedMethods.Add(GetNameOrDefault(MethodName.OperatorNotEquals, classMethods, KnownMethods.IsOperatorNotEquals));
+                implementedMethods.Add(GetNameOrDefault(MethodName.ObjectEquals, classMethods, KnownMethods.IsObjectEquals));
+                implementedMethods.Add(GetNameOrDefault(MethodName.ObjectGetHashCode, classMethods, KnownMethods.IsObjectGetHashCode));
+                implementedMethods.Remove(null);
+
+                var requiredMethods = new HashSet<string>();
+
+                if (implementedMethods.Contains(MethodName.OperatorPlus))
```

Comment:
I would create a method `GetMissingMethods(IEnumerable<string> implementedMethods)` doing the following ifs and returning the result of the except.